File: sdk/reproit-backend-dotnet/ReproitBackend/Replay.cs
// Hermetic replay for reproit-backend-dotnet.
//
// When `REPROIT_REPLAY` names a `reproit-backend-capture` payload, the same boundary that
// records exchanges at capture time SERVES them instead, so the application re-executes
// against exactly what production saw with no live dependency at all.
//
// Determinism is a contract here, not a similarity score. Matching is strict per-operation
// ordinals: within one operation (method plus path for HTTP, statement text for pg) exchanges
// are consumed in recorded order, so pooled db clients and LLM tool-call loops that interleave
// operations still match exactly. Recorded `$reproit` redaction placeholders match any value
// at their position, and a body truncated at capture fails closed. The first unmatched call is
// a DIVERGENCE, reported as a structured `REPROIT:DIVERGENCE` stderr line, byte-identical to
// the Node SDK's (insertion-order compact JSON, Json.Compact), with a `bodyDelta` naming
// WHERE the bodies differ: chat-shaped bodies name the first differing message index, unknown
// shapes fall back to the byte offset of the first differing byte.
//
// The envelope pins the replay: the time zone comes from the capture and Rng() yields the
// seeded stream. Honesty note: the seed makes REPLAY runs deterministic; it does not
// reproduce the randomness the app drew in production.

using System.Globalization;

namespace ReproitBackend;

public sealed class Replay
{
    // The structured divergence marker, byte-identical to the Node SDK's.
    public const string DivergenceMarker = "REPROIT:DIVERGENCE ";

    private sealed class Entry
    {
        public required Dictionary<string, object?> Exchange { get; init; }
        public bool Consumed { get; set; }
    }

    private readonly Dictionary<string, object?> _envelope;
    private readonly List<Entry> _exchanges = new();
    private readonly object _lock = new();

    private Replay(Dictionary<string, object?> envelope, List<Dictionary<string, object?>> found)
    {
        _envelope = envelope;
        foreach (var exchange in found) _exchanges.Add(new Entry { Exchange = exchange });
    }

    // Load the capture named by `path`, or null when it is not a supported
    // `reproit-backend-capture` payload. Never throws: replay mode simply does not engage.
    public static Replay? Load(string path)
    {
        try
        {
            if (Json.Parse(File.ReadAllText(path)) is not Dictionary<string, object?> payload)
            {
                return null;
            }
            if (payload.GetValueOrDefault("format") as string != "reproit-backend-capture")
            {
                return null;
            }
            var version = payload.GetValueOrDefault("version") switch
            {
                long value => value,
                int value => value,
                double value => (long)value,
                _ => 0L,
            };
            if (version is < 1 or > 2) return null;
            var found = new List<Dictionary<string, object?>>();
            if (payload.GetValueOrDefault("events") is List<object?> events)
            {
                foreach (var item in events)
                {
                    if (item is not Dictionary<string, object?> evt) continue;
                    if (evt.GetValueOrDefault("kind") as string != "effect") continue;
                    if (evt.GetValueOrDefault("exchange") is Dictionary<string, object?> exchange)
                    {
                        found.Add(exchange);
                    }
                }
            }
            var envelope = payload.GetValueOrDefault("envelope") as Dictionary<string, object?>
                ?? new Dictionary<string, object?>();
            return new Replay(envelope, found);
        }
        catch (Exception)
        {
            return null;
        }
    }

    // The capture's time zone, resolved from the envelope.
    public TimeZoneInfo? PinnedTimeZone()
    {
        if (_envelope.GetValueOrDefault("tz") is not string tz || tz.Length == 0) return null;
        try
        {
            return TimeZoneInfo.FindSystemTimeZoneById(tz);
        }
        catch (Exception)
        {
            return null;
        }
    }

    // Pin the process time zone to the capture's, so replayed code reading
    // DateTime.Now or TimeZoneInfo.Local sees what production saw.
    //
    // On Unix .NET resolves TimeZoneInfo.Local from the TZ environment
    // variable, and ClearCachedData drops the cached Local so the next read
    // re-resolves it. Windows resolves the local zone from the registry and
    // ignores TZ, so there the zone cannot be pinned and PinnedTimeZone stays
    // the app-readable fallback. Returns true only when the zone was really
    // pinned, so no caller can mistake the fallback for a pin.
    public bool PinTimeZone()
    {
        if (_envelope.GetValueOrDefault("tz") is not string tz || tz.Length == 0) return false;
        if (OperatingSystem.IsWindows()) return false;
        try
        {
            Environment.SetEnvironmentVariable("TZ", tz);
            TimeZoneInfo.ClearCachedData();
            return TimeZoneInfo.Local.Id == tz;
        }
        catch (Exception)
        {
            return false;
        }
    }

    // The capture moment from the envelope, or null when the capture carries none.
    public long? ObservedAtMs() => _envelope.GetValueOrDefault("observedAtMs") switch
    {
        long value => value,
        int value => value,
        double value => (long)value,
        _ => null,
    };

    // Deterministic xorshift64* stream from the capture's `replaySeed`, or null when the
    // capture carries no envelope seed.
    public ReplayRng? Rng()
    {
        if (_envelope.GetValueOrDefault("replaySeed") is not string seed || seed.Length == 0)
        {
            return null;
        }
        var hex = seed.Length > 16 ? seed[..16] : seed;
        return ulong.TryParse(hex, NumberStyles.HexNumber, CultureInfo.InvariantCulture,
            out var state) ? new ReplayRng(state | 1UL) : null;
    }

    // Strict per-operation ordinal match. Within one operation (the key below) the next
    // unconsumed exchange is the ONLY candidate; skipping it silently would be a fuzzy match.
    // Other operations' exchanges may interleave (db pooling, tool-call loops), which is why
    // the key filters first. Null is a divergence, reported.
    internal Dictionary<string, object?>? Matched(
        string protocol, Dictionary<string, object?> probe)
    {
        var key = OperationKey(protocol, probe);
        lock (_lock)
        {
            foreach (var entry in _exchanges)
            {
                if (entry.Consumed) continue;
                if (entry.Exchange.GetValueOrDefault("protocol") as string != protocol) continue;
                var recorded =
                    entry.Exchange.GetValueOrDefault("request") as Dictionary<string, object?>
                    ?? new Dictionary<string, object?>();
                if (OperationKey(protocol, recorded) != key) continue;
                var hit = protocol == "http"
                    ? HttpMatches(recorded, probe)
                    : DbMatches(recorded, probe);
                if (hit)
                {
                    entry.Consumed = true;
                    return entry.Exchange;
                }
                break;
            }
        }
        Diverge(protocol, probe);
        return null;
    }

    // One operation's identity for ordinal matching: HTTP is method plus path and query, pg
    // is the exact statement text.
    internal static string OperationKey(string protocol, Dictionary<string, object?> request) =>
        protocol == "http"
            ? (request.GetValueOrDefault("method") as string ?? string.Empty) + " " +
                PathAndQuery(request.GetValueOrDefault("url"))
            : request.GetValueOrDefault("text") as string ?? string.Empty;

    // Report a divergence on stderr in the shared structured shape. Field order mirrors the
    // Node reference so the marker line is byte-comparable across SDKs; compact insertion-
    // order separators for the same reason.
    internal void Diverge(string protocol, Dictionary<string, object?> probe)
    {
        var key = OperationKey(protocol, probe);
        Dictionary<string, object?>? expected = null;
        Dictionary<string, object?>? firstCandidate = null;
        long consumed = 0;
        long total;
        lock (_lock)
        {
            foreach (var entry in _exchanges)
            {
                if (entry.Consumed)
                {
                    consumed += 1;
                    continue;
                }
                if (entry.Exchange.GetValueOrDefault("protocol") as string != protocol)
                {
                    continue;
                }
                var request =
                    entry.Exchange.GetValueOrDefault("request") as Dictionary<string, object?>
                    ?? new Dictionary<string, object?>();
                firstCandidate ??= request;
                if (expected == null && OperationKey(protocol, request) == key)
                {
                    expected = request;
                }
            }
            total = _exchanges.Count;
        }
        expected ??= firstCandidate;
        var report = new Dictionary<string, object?>
        {
            ["protocol"] = protocol,
            ["got"] = probe,
            ["expected"] = expected,
            ["consumed"] = consumed,
            ["total"] = total,
        };
        // Prompt drift: when the recorded and live bodies both exist and differ, name WHERE.
        var delta = expected == null
            ? null
            : BodyDelta(
                expected.TryGetValue("body", out var recordedBody) ? recordedBody : Absent,
                probe.TryGetValue("body", out var liveBody) ? liveBody : Absent);
        if (delta != null) report["bodyDelta"] = delta;
        Console.Error.WriteLine(DivergenceMarker + Json.Compact(report));
    }

    // Distinct from null: an ABSENT body is "the request had no body key", which the delta
    // must not confuse with an explicit JSON null.
    internal static readonly object Absent = new();

    // The messages array of an OpenAI/Anthropic-shaped chat body, else null.
    private static List<object?>? ChatMessages(object? body) =>
        body is Dictionary<string, object?> map &&
        map.GetValueOrDefault("messages") is List<object?> messages ? messages : null;

    private static byte[] DeltaBytes(object? value) =>
        System.Text.Encoding.UTF8.GetBytes(value is string text ? text : Json.Compact(value));

    // Locate the first difference between a recorded request body and a live one, modulo
    // redaction placeholders. Null when there is nothing to report (either body missing, or
    // no difference the matcher would object to).
    internal static Dictionary<string, object?>? BodyDelta(object? recorded, object? live)
    {
        if (ReferenceEquals(recorded, Absent) || ReferenceEquals(live, Absent)) return null;
        if (Matches(recorded, live)) return null;
        var recordedMessages = ChatMessages(recorded);
        var liveMessages = ChatMessages(live);
        if (recordedMessages != null && liveMessages != null)
        {
            var bound = Math.Min(recordedMessages.Count, liveMessages.Count);
            int? index = null;
            for (var i = 0; i < bound; i++)
            {
                if (!Matches(recordedMessages[i], liveMessages[i]))
                {
                    index = i;
                    break;
                }
            }
            // All shared indexes match: the drift is a longer or shorter conversation, and
            // the first differing message is the first unshared one. If lengths also agree
            // the drift is outside `messages`; fall through to bytes.
            if (index == null && recordedMessages.Count != liveMessages.Count) index = bound;
            if (index != null)
            {
                return new Dictionary<string, object?>
                {
                    ["kind"] = "message",
                    ["firstDifferingMessage"] = (long)index.Value,
                    ["recordedMessages"] = (long)recordedMessages.Count,
                    ["liveMessages"] = (long)liveMessages.Count,
                };
            }
        }
        var recordedBytes = DeltaBytes(recorded);
        var liveBytes = DeltaBytes(live);
        var length = Math.Min(recordedBytes.Length, liveBytes.Length);
        long offset = length;
        for (var i = 0; i < length; i++)
        {
            if (recordedBytes[i] != liveBytes[i])
            {
                offset = i;
                break;
            }
        }
        return new Dictionary<string, object?> { ["kind"] = "byte", ["offset"] = offset };
    }

    // The served shape of one HTTP probe: what the instrumented handler synthesizes a
    // response from. `Chunks` is present only for a recorded stream shape.
    public sealed class Served
    {
        public required int Status { get; init; }
        public required Dictionary<string, object?> Headers { get; init; }
        public required string BodyText { get; init; }
        public List<byte[]>? Chunks { get; init; }
    }

    // Resolve a live HTTP probe against the session, entirely in process (no sockets). A
    // divergence, a body truncated at capture, and truncated stream boundaries all serve a
    // hard 599 so the application observes an attributable failure instead of a guess.
    public Served ServeHttp(Dictionary<string, object?> probe)
    {
        var recorded = Matched("http", probe);
        if (recorded == null) return Diverged599("diverged");
        var response = recorded.GetValueOrDefault("response") as Dictionary<string, object?>
            ?? new Dictionary<string, object?>();
        if (response.GetValueOrDefault("truncated") as bool? == true)
        {
            // The capture kept identity but not bytes; serving a guessed body would be a
            // silent lie. Fail closed with the named reason.
            var diverged = new Dictionary<string, object?>(probe) { ["truncated"] = true };
            Diverge("http", diverged);
            return Diverged599("truncated-exchange-body");
        }
        var headers = new Dictionary<string, object?>();
        if (response.GetValueOrDefault("headers") is Dictionary<string, object?> recordedHeaders)
        {
            foreach (var (name, value) in recordedHeaders)
            {
                var lower = name.ToLowerInvariant();
                if (lower is "content-length" or "transfer-encoding" or "content-encoding")
                {
                    continue;
                }
                headers[name] = value;
            }
        }
        var body = response.GetValueOrDefault("body");
        var bodyText = body switch
        {
            null => string.Empty,
            string text => text,
            // Insertion-order compact: byte-identical to the Node reference's
            // JSON.stringify of the same recorded body.
            _ => Json.Compact(body),
        };
        var status = response.GetValueOrDefault("status") switch
        {
            long value => (int)value,
            int value => value,
            double value => (int)value,
            _ => 200,
        };
        List<byte[]>? chunks = null;
        if (response.GetValueOrDefault("stream") is Dictionary<string, object?> stream &&
            stream.GetValueOrDefault("chunks") is List<object?> boundaries)
        {
            if (stream.GetValueOrDefault("truncated") as bool? == true)
            {
                // The capture kept the body but not every chunk boundary; serving a guessed
                // stream shape would be a silent lie. Fail closed, named.
                var diverged = new Dictionary<string, object?>(probe)
                {
                    ["streamBoundariesTruncated"] = true,
                };
                Diverge("http", diverged);
                return Diverged599("truncated-stream-boundaries");
            }
            chunks = SplitChunks(bodyText, boundaries);
        }
        return new Served
        {
            Status = status,
            Headers = headers,
            BodyText = bodyText,
            Chunks = chunks,
        };
    }

    // Split a replayed body at the recorded chunk boundaries (byte lengths). Redaction can
    // change body byte counts, so lengths are clamped and the last chunk absorbs any
    // remainder: the CHUNK COUNT (the stream shape the app observed) is preserved exactly,
    // the recorded content never padded.
    internal static List<byte[]> SplitChunks(string bodyText, List<object?> lengths)
    {
        var raw = System.Text.Encoding.UTF8.GetBytes(bodyText);
        var chunks = new List<byte[]>(lengths.Count);
        var offset = 0;
        for (var index = 0; index < lengths.Count; index++)
        {
            var last = index == lengths.Count - 1;
            var size = lengths[index] switch
            {
                long value when value > 0 => (int)value,
                int value when value > 0 => value,
                double value when value > 0 => (int)value,
                _ => 0,
            };
            var end = last ? raw.Length : Math.Min(offset + size, raw.Length);
            chunks.Add(raw[offset..end]);
            offset = end;
        }
        return chunks;
    }

    private static Served Diverged599(string reason) => new()
    {
        Status = 599,
        Headers = new Dictionary<string, object?> { ["content-type"] = "application/json" },
        BodyText = Json.Compact(new Dictionary<string, object?> { ["reproit"] = reason }),
    };

    // Parse a request body the way the capture recorded it: declared JSON parses (so the
    // matcher compares structure), everything else stays text.
    public static object? TryJson(string text, string? contentType)
    {
        if (contentType != null &&
            contentType.Contains("application/json", StringComparison.Ordinal))
        {
            try
            {
                return Json.Parse(text);
            }
            catch (Exception)
            {
                return text;
            }
        }
        return text;
    }

    // Method, path and query of the original URL, and body modulo placeholders. Recorded
    // headers are deliberately not matched: they carry per-run noise that would turn every
    // replay into a divergence.
    internal static bool HttpMatches(
        Dictionary<string, object?> recorded, Dictionary<string, object?> probe)
    {
        if (!Equals(recorded.GetValueOrDefault("method"), probe.GetValueOrDefault("method")))
        {
            return false;
        }
        if (PathAndQuery(recorded.GetValueOrDefault("url")) !=
            PathAndQuery(probe.GetValueOrDefault("url")))
        {
            return false;
        }
        return !recorded.ContainsKey("body")
            || Matches(recorded.GetValueOrDefault("body"), probe.GetValueOrDefault("body"));
    }

    // Exact statement text, values modulo placeholders.
    internal static bool DbMatches(
        Dictionary<string, object?> recorded, Dictionary<string, object?> probe)
    {
        if (!Equals(recorded.GetValueOrDefault("text"), probe.GetValueOrDefault("text")))
        {
            return false;
        }
        return !recorded.ContainsKey("values")
            || Matches(recorded.GetValueOrDefault("values"), probe.GetValueOrDefault("values"));
    }

    internal static string PathAndQuery(object? url)
    {
        if (url is not string text) return string.Empty;
        return Uri.TryCreate(text, UriKind.Absolute, out var parsed)
            ? parsed.PathAndQuery : text;
    }

    // A recorded value matches a live one when equal, or when the recorded side is a
    // `$reproit` redaction placeholder (any value stood here at capture). Objects compare per
    // key; a recorded null matches anything.
    internal static bool Matches(object? recorded, object? live)
    {
        switch (recorded)
        {
            case null:
                return true;
            case Dictionary<string, object?> map:
                if (map.ContainsKey("$reproit")) return true;
                if (live is not Dictionary<string, object?> liveMap) return false;
                return map.All(entry =>
                    Matches(entry.Value, liveMap.GetValueOrDefault(entry.Key)));
            case List<object?> list:
                if (live is not List<object?> liveList || liveList.Count != list.Count)
                {
                    return false;
                }
                return !list.Where((item, index) => !Matches(item, liveList[index])).Any();
            default:
                if (recorded is IConvertible && live is IConvertible
                    && recorded is not string && live is not string)
                {
                    return Convert.ToDouble(recorded, CultureInfo.InvariantCulture)
                        .Equals(Convert.ToDouble(live, CultureInfo.InvariantCulture));
                }
                return recorded.Equals(live);
        }
    }
}

// The seeded replay stream; matches the Node SDK's draw shape.
public sealed class ReplayRng
{
    private ulong _state;

    internal ReplayRng(ulong state)
    {
        _state = state;
    }

    // The next draw in [0, 1).
    public double NextDouble() => (NextUInt64() >> 11) / (double)(1UL << 53);

    // The next raw 64-bit word of the stream. NextDouble scales one of these, so the two
    // draws share one sequence; crypto-random byte draws pull whole words from it.
    internal ulong NextUInt64()
    {
        _state ^= _state << 13;
        _state ^= _state >> 7;
        _state ^= _state << 17;
        return unchecked(_state * 0x2545f4914f6cdd1dUL);
    }
}

File: sdk/reproit-backend-rb/lib/reproit_backend_rb/replay.rb
# Hermetic replay mode for reproit-backend-rb.
#
# When `REPROIT_REPLAY` names a `reproit-backend-capture` payload, the same
# Net::HTTP hook and database helper that record exchanges at capture time
# SERVE them instead, so the application re-executes against exactly what
# production saw with no live dependency.
#
# Determinism is a contract here, not a similarity score. Matching is strict
# per-operation ordinals: within one operation (method plus path for HTTP,
# statement text for the database) exchanges are consumed in recorded order,
# so pooled database clients and LLM tool-call loops that interleave
# operations still match exactly. Recorded `$reproit` redaction placeholders
# match any value at their position; nothing else is tolerated. The first
# unmatched call is a DIVERGENCE: it writes the structured
# `REPROIT:DIVERGENCE ` line to stderr (with a `bodyDelta` naming WHERE the
# bodies differ; chat-shaped bodies name the first differing message index)
# and the call fails with status 599 (HTTP) or a raised error (database),
# never a fuzzy match.
#
# The envelope pins replay determinism: `TZ`, the wall clock, `srand` for
# Kernel#rand, and `replay_rng` yields the cross-SDK seeded stream. Honesty
# note: the seed makes REPLAY runs deterministic; it does not reproduce the
# randomness the app drew in production. SecureRandom reads OS entropy
# directly and CANNOT be pinned; that is a named gap, not a downgrade.

require "json"
require "uri"

module ReproitBackendRb
  module Replay
    DIVERGENCE_MARKER = "REPROIT:DIVERGENCE "
    # Sentinel for a body FIELD that is absent, as distinct from a recorded
    # null body (the Node reference's `undefined` vs `null`). A bodyDelta is
    # only computed when both sides actually carry a body field.
    ABSENT = Object.new.freeze

    # Deterministic xorshift64* stream, identical to the Node and Rust SDKs.
    class Rng
      def initialize(state)
        @state = state | 1
      end

      MASK = 0xFFFF_FFFF_FFFF_FFFF

      def next_float
        @state ^= (@state << 13) & MASK
        @state ^= @state >> 7
        @state ^= (@state << 17) & MASK
        mixed = (@state * 0x2545_F491_4F6C_DD1D) & MASK
        (mixed >> 11).to_f / (1 << 53).to_f
      end
    end

    class Session
      attr_reader :envelope

      def self.load(path)
        payload = JSON.parse(File.read(path))
        unless payload["format"] == "reproit-backend-capture"
          raise ArgumentError, "REPROIT_REPLAY file is not a reproit-backend-capture payload"
        end
        version = payload["version"]
        unless version.is_a?(Integer) && version >= 1 && version <= 2
          raise ArgumentError, "unsupported capture version #{version.inspect}"
        end
        new(payload)
      end

      def initialize(payload)
        @envelope = payload["envelope"]
        @entries = (payload["events"] || []).filter_map do |event|
          next unless event.is_a?(Hash) && event["kind"] == "effect" && event["exchange"]
          { "exchange" => event["exchange"], "consumed" => false }
        end
        @diverged = false
        @lock = Mutex.new
      end

      def diverged?
        @lock.synchronize { @diverged }
      end

      # Strict per-operation ordinal match. Returns the exchange or nil
      # (divergence).
      def match(protocol, probe)
        key = Replay.operation_key(protocol, probe)
        found = @lock.synchronize do
          hit = nil
          @entries.each do |candidate|
            next if candidate["consumed"] || candidate["exchange"]["protocol"] != protocol
            request = candidate["exchange"]["request"] || {}
            next unless Replay.operation_key(protocol, request) == key
            if Replay.request_matches?(protocol, request, probe)
              candidate["consumed"] = true
              hit = candidate["exchange"]
            end
            # Strict ordinal within an operation: the next unconsumed
            # exchange of THIS operation is the only candidate; skipping it
            # silently would be a fuzzy match. Other operations' exchanges
            # may interleave (database pooling, tool-call loops), which is
            # why the key filters above.
            break
          end
          hit
        end
        return found if found
        diverge(protocol, probe)
        nil
      end

      def diverge(protocol, probe)
        key = Replay.operation_key(protocol, probe)
        report = @lock.synchronize do
          @diverged = true
          candidates = @entries.reject do |candidate|
            candidate["consumed"] || candidate["exchange"]["protocol"] != protocol
          end
          expected = candidates.find do |candidate|
            Replay.operation_key(protocol, candidate["exchange"]["request"] || {}) == key
          end || candidates.first
          # Field insertion order mirrors the Node reference so the marker
          # line is byte-comparable across SDKs; JSON.generate is compact.
          lines = {
            "protocol" => protocol,
            "got" => probe,
            "expected" => expected.nil? ? nil : expected["exchange"]["request"],
            "consumed" => @entries.count { |candidate| candidate["consumed"] },
            "total" => @entries.length,
          }
          # Prompt drift: when the recorded and live bodies both exist and
          # differ, name WHERE they differ. Chat-shaped bodies (OpenAI or
          # Anthropic messages arrays) name the first differing message
          # index; unknown shapes fall back to the first differing byte.
          unless expected.nil?
            recorded_body = (expected["exchange"]["request"] || {}).fetch("body", ABSENT)
            delta = Replay.body_delta(recorded_body, probe.fetch("body", ABSENT))
            lines["bodyDelta"] = delta unless delta.nil?
          end
          lines
        end
        # Written raw, never through `warn`: Kernel#warn with `uplevel`
        # prepends "file:line: warning: ", which would break the marker
        # prefix the CLI matches on. The line must be byte-identical to the
        # Node and Rust SDKs' so one parser reads every platform.
        $stderr.write(DIVERGENCE_MARKER + JSON.generate(report) + "\n")
        nil
      end
    end

    module_function

    # One operation's identity for ordinal matching: HTTP is method plus
    # path and query, database is the exact statement text.
    def operation_key(protocol, request)
      if protocol == "http"
        request["method"].to_s + " " + path_and_query(request["url"])
      else
        request["text"].to_s
      end
    end

    # The messages array of an OpenAI/Anthropic-shaped chat body, else nil.
    def chat_messages(body)
      body.is_a?(Hash) && body["messages"].is_a?(Array) ? body["messages"] : nil
    end

    def delta_bytes(value)
      raw = value.is_a?(String) ? value : JSON.generate(value)
      raw.dup.force_encoding(Encoding::BINARY)
    end

    # Locate the first difference between a recorded request body and a live
    # one, modulo redaction placeholders. Nil when there is nothing to
    # report (either body missing, or no difference the matcher objects to).
    def body_delta(recorded, live)
      return nil if recorded.equal?(ABSENT) || live.equal?(ABSENT)
      return nil if matches?(recorded, live)
      recorded_messages = chat_messages(recorded)
      live_messages = chat_messages(live)
      if recorded_messages && live_messages
        bound = [recorded_messages.length, live_messages.length].min
        index = (0...bound).find do |i|
          !matches?(recorded_messages[i], live_messages[i])
        end
        # All shared indexes match: the drift is a longer or shorter
        # conversation, and the first differing message is the first
        # unshared one. If lengths also agree the drift is outside
        # `messages`; fall through to bytes.
        if index.nil? && recorded_messages.length != live_messages.length
          index = bound
        end
        unless index.nil?
          return {
            "kind" => "message",
            "firstDifferingMessage" => index,
            "recordedMessages" => recorded_messages.length,
            "liveMessages" => live_messages.length,
          }
        end
      end
      recorded_raw = delta_bytes(recorded)
      live_raw = delta_bytes(live)
      bound = [recorded_raw.bytesize, live_raw.bytesize].min
      offset = (0...bound).find { |i| recorded_raw.getbyte(i) != live_raw.getbyte(i) } || bound
      { "kind" => "byte", "offset" => offset }
    end

    # A recorded value matches a live one when equal, when the recorded side
    # is a `$reproit` redaction placeholder (any value stood there at
    # capture), or when the recorded side is absent. Hashes compare per key.
    def matches?(recorded, live)
      case recorded
      when nil then true
      when Hash
        return true if recorded.key?("$reproit")
        return false unless live.is_a?(Hash)
        recorded.all? { |key, value| matches?(value, live[key]) }
      when Array
        return false unless live.is_a?(Array) && live.length == recorded.length
        recorded.each_with_index.all? { |item, index| matches?(item, live[index]) }
      else recorded == live
      end
    end

    def request_matches?(protocol, recorded, probe)
      if protocol == "http"
        return false unless recorded["method"] == probe["method"]
        return false unless path_and_query(recorded["url"]) == path_and_query(probe["url"])
        # Recorded headers are deliberately not matched: they carry per-run
        # noise (dates, connection management) that would turn every replay
        # into a divergence.
        matches?(recorded["body"], probe["body"])
      else
        return false unless recorded["text"] == probe["text"]
        matches?(recorded["values"], probe["values"])
      end
    end

    def path_and_query(url)
      parsed = URI.parse(url.to_s)
      query = parsed.query.nil? ? "" : "?" + parsed.query
      (parsed.path.nil? || parsed.path.empty? ? "/" : parsed.path) + query
    rescue URI::InvalidURIError
      url.to_s
    end

    # Resolve a live HTTP probe against the session entirely in process. A
    # divergence and a truncated-at-capture body both serve a hard 599 so the
    # application observes an attributable failure instead of a guess.
    def serve_http(session, probe)
      recorded = session.match("http", probe)
      return diverged_599("diverged") if recorded.nil?
      response = recorded["response"] || {}
      if response["truncated"] == true
        # The capture kept identity but not bytes; serving a guessed body
        # would be a silent lie. Fail closed with the named reason.
        session.diverge("http", probe.merge("truncated" => true))
        return diverged_599("truncated-exchange-body")
      end
      headers = (response["headers"] || {}).reject do |name, _|
        %w[content-length transfer-encoding content-encoding].include?(name.to_s.downcase)
      end
      body = response["body"]
      text = case body
             when nil then ""
             when String then body
             else JSON.generate(body)
             end
      served = { "status" => response["status"] || 200, "headers" => headers, "body" => text }
      stream = response["stream"]
      if stream.is_a?(Hash) && stream["chunks"].is_a?(Array)
        if stream["truncated"] == true
          # The capture kept the body but not every chunk boundary; serving
          # a guessed stream shape would be a silent lie. Fail closed with
          # the named reason.
          session.diverge("http", probe.merge("streamBoundariesTruncated" => true))
          return diverged_599("truncated-stream-boundaries")
        end
        served["chunks"] = split_chunks(text, stream["chunks"])
      end
      served
    end

    # Split a replayed body at the recorded chunk boundaries (byte lengths).
    # Redaction can change body byte counts, so lengths are clamped and the
    # last chunk absorbs any remainder: the CHUNK COUNT (the stream shape the
    # app observed) is preserved exactly, the recorded content never padded.
    def split_chunks(body_text, lengths)
      raw = body_text.dup.force_encoding(Encoding::BINARY)
      chunks = []
      offset = 0
      lengths.each_with_index do |length, index|
        last = index == lengths.length - 1
        size = length.is_a?(Integer) && length.positive? ? length : 0
        finish = last ? raw.bytesize : [offset + size, raw.bytesize].min
        chunks << raw.byteslice(offset, finish - offset)
        offset = finish
      end
      chunks
    end

    def diverged_599(reason)
      {
        "status" => 599,
        "headers" => { "content-type" => "application/json" },
        "body" => JSON.generate({ "reproit" => reason }),
      }
    end

    def try_json(text, content_type)
      return text unless content_type.to_s.include?("application/json")
      JSON.parse(text)
    rescue JSON::ParserError
      text
    end

    # Pin process determinism from the capture envelope: the time zone, the
    # wall clock, and the seeded stream.
    #
    # The clock is anchored by prepending a module to Time's singleton class
    # (the Timecop pattern), which is the one safe process-wide hook Ruby
    # offers. It runs ONLY under REPROIT_REPLAY. Like the Node reference this
    # OFFSETS rather than freezes: replayed code sees the capture's instant
    # and still observes elapsed time within the run, so a timeout loop
    # terminates instead of hanging.
    #
    # Honesty note: this makes replay runs repeatable. It does not reproduce
    # the exact instants production observed between events.
    def pin_envelope(envelope)
      return unless envelope.is_a?(Hash)
      tz = envelope["tz"]
      ENV["TZ"] = tz if tz.is_a?(String) && !tz.empty?
      pin_clock(envelope["observedAtMs"])
      pin_rand(envelope["replaySeed"])
    end

    # Seed Kernel#rand (the default Random stream) from the envelope so
    # replayed application draws are repeatable across runs. `replay_rng`
    # additionally exposes the cross-SDK xorshift64* stream. SecureRandom
    # reads OS entropy directly and cannot be pinned: a NAMED gap.
    def pin_rand(seed)
      return unless seed.is_a?(String) && !seed.empty?
      srand(seed[0, 16].ljust(16, "0").to_i(16))
      true
    end

    def pin_clock(observed_at_ms)
      return unless observed_at_ms.is_a?(Numeric)
      offset = (observed_at_ms / 1000.0) - Time.now.to_f
      Time.singleton_class.prepend(clock_module(offset))
      Process.singleton_class.prepend(process_clock_module(offset))
      true
    end

    def clock_module(offset)
      Module.new do
        define_method(:now) { |*args| super(*args) + offset }
        define_method(:reproit_clock_offset) { offset }
      end
    end

    def process_clock_module(offset)
      Module.new do
        define_method(:clock_gettime) do |clock_id, *rest|
          value = super(clock_id, *rest)
          # Only the wall clock is anchored; monotonic clocks must keep
          # advancing from their own epoch or duration math breaks.
          next value unless clock_id == Process::CLOCK_REALTIME
          unit = rest.first
          case unit
          when :millisecond then value + (offset * 1000).round
          when :microsecond then value + (offset * 1_000_000).round
          when :nanosecond then value + (offset * 1_000_000_000).round
          else value + offset
          end
        end
      end
    end

    def rng_for(envelope)
      return nil unless envelope.is_a?(Hash)
      seed = envelope["replaySeed"]
      return nil unless seed.is_a?(String) && !seed.empty?
      Rng.new(seed[0, 16].ljust(16, "0").to_i(16))
    end
  end
end
